velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,243 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/prestosql/CheckedArithmeticImpl.h"
+#include "velox/type/DecimalUtils.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <typename R, typename A, typename B, typename Operation>
+class DecimalBaseFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    uint8_t aRescale, bRescale, aScale, bScale, aPrecision, bPrecision,
+        rPrecision, rScale;
+    getDecimalPrecisionScale(args[0]->type(), aPrecision, aScale);
+    getDecimalPrecisionScale(args[1]->type(), bPrecision, bScale);
+    getDecimalPrecisionScale(resultType, rPrecision, rScale);
+    aRescale = Operation::computeRescaleFactor(aScale, bScale, rScale);
+    bRescale = Operation::computeRescaleFactor(bScale, aScale, rScale);
+    if (args[0]->isConstantEncoding() &&
+        args[1]->encoding() == VectorEncoding::Simple::FLAT) {
+      // Fast path for (flat, const).
+      auto constant = args[0]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+      auto flatValues = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults =
+          getRawResults(rows, args[1], resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], constant, rawValues[row], aRescale, bRescale);
+      });
+    } else if (
+        args[0]->encoding() == VectorEncoding::Simple::FLAT &&
+        args[1]->isConstantEncoding()) {
+      auto flatValues = args[0]->asUnchecked<FlatVector<A>>();
+      auto constant = args[1]->asUnchecked<SimpleVector<B>>()->valueAt(0);
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults =
+          getRawResults(rows, args[1], resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawValues[row], constant, aRescale, bRescale);
+      });
+    } else if (
+        args[0]->encoding() == VectorEncoding::Simple::FLAT &&
+        args[1]->encoding() == VectorEncoding::Simple::FLAT) {
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawB = flatB->mutableRawValues();
+
+      R* rawResults = nullptr;
+      if (!(*result)) {
+        if (resultType->kind() == flatA->typeKind() &&
+            BaseVector::isReusableFlatVector(args[0])) {
+          rawResults = (R*)rawA;
```

Comment:
use `static_cast`.

Reply:
"Static_cast from 'facebook::velox::ShortDecimal *' to 'facebook::velox::LongDecimal *', which are not related by inheritance, is not allowed"

Reply:
use `reinterpret_cast` then.